### PR TITLE
Fix DC for batch_size>1

### PR DESCRIPTION
**Pull request by chaithyagr**
Currently, the NCPDNet works only for batch_size = 1. 

This is a minute bug.

### CODE REVIEW

File: fastmri_recon/models/utils/fourier.py [hunk 1]
```diff
@@ -255,7 +255,7 @@ def adj_op(self, inputs):
             else:
                 kspace, ktraj, shape = inputs
         if self.density_compensation:
-            kspace = tf.cast(dcomp, kspace.dtype) * kspace[..., 0]
+            kspace = tf.cast(dcomp[:, None], kspace.dtype) * kspace[..., 0]
```
Comment (zaccharieramzi) [inline fastmri_recon/models/utils/fourier.py:258]: did you test this for both multicoil and singlecoil ?

I indeed was using a batch size of one even when testing..., my bad

Comment (chaithyagr) [reply fastmri_recon/models/utils/fourier.py:258]: We test a version of this in the tf-fastmri-data. 
I have not tested for single coil.. I am not sure how extensive your testing is for this..

**Comment by chaithyagr** [comment]
I have also updated this branch with a method to save optimizer weights at model checkpoint...
This is particularly useful in case the run crashes as then we will have last model and its associated weights.

File: fastmri_recon/training_scripts/model_saving_workaround.py [hunk 2]
```diff
@@ -1,9 +1,30 @@
...
 from tensorflow.keras.callbacks import ModelCheckpoint
+import tensorflow.keras.backend as K
+
 
 # taken from https://github.com/tensorflow/tensorflow/issues/39679
 class ModelCheckpointWorkAround(ModelCheckpoint):
+    def __init__(self, filepath, save_optimizer=True, **kwargs):
+        os.makedirs(os.path.dirname(filepath), exist_ok=True)
+        self.save_optimizer = save_optimizer
+        super(ModelCheckpointWithOptimizer, self).__init__(filepath=filepath, **kwargs)
```
Comment (zaccharieramzi) [inline fastmri_recon/training_scripts/model_saving_workaround.py:13]: Watch out the naming

Comment (zaccharieramzi) [reply fastmri_recon/training_scripts/model_saving_workaround.py:13]: Also why do you need the filepath logic?

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
I don't think the checkpoint stuff should go in this PR

File: fastmri_recon/models/utils/fourier.py (hunk 1 above)
Comment (chaithyagr) [reply fastmri_recon/models/utils/fourier.py:258]: Adding here that I tested it even for single coil

File: fastmri_recon/training_scripts/model_saving_workaround.py (hunk 2 above)
Comment (chaithyagr) [reply fastmri_recon/training_scripts/model_saving_workaround.py:13]: Well, I had codes failing just cause we didnt have the directory present :\. This is a good backup, anyway it just creates the output directory if needed, and if it exists, would not do anything.

Comment (zaccharieramzi) [reply fastmri_recon/training_scripts/model_saving_workaround.py:13]:
```suggestion
        super(ModelCheckpointWorkAround, self).__init__(filepath=filepath, **kwargs)
```

File: fastmri_recon/training_scripts/model_saving_workaround.py [hunk 3]
```diff
@@ -1,9 +1,30 @@
...
+
     def set_model(self, model):
         # Work around, so that the if at
         # https://github.com/tensorflow/tensorflow/blob/1186e3f2098793952aa82bf356dfe51b967fb26c/tensorflow/python/keras/callbacks.py#L1189
         # is skipped, so that self.save_weights_only remains False.
         self.model = model
+
+    def _save_model(self, epoch, logs):
+        # Save the model with super
+        super(ModelCheckpointWithOptimizer, self)._save_model(epoch, logs)
```
Comment (zaccharieramzi) [inline fastmri_recon/training_scripts/model_saving_workaround.py:23]:
```suggestion
        super(ModelCheckpointWorkAround, self)._save_model(epoch, logs)
```

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]
The tests are failing because of the new model checkpointer

**Review by zaccharieramzi — APPROVED** [review]

**Pull request merged**